### ISSUE TITLE
FID reaches to 200 during training

### ISSUE THREAD

**Issue by sydney0zq** [saic-mdal/lama#43]
Hi, do you have trained the model to ensure the accuracy is promised? I have train the lama-fourier with 4 gpus(bs = 10), each epoch costs me 6 hours and it needs 10 days to finished. Could you please share your training log as a reference. The log pasted in README.md is half-baked. And my FID even reaches to 200 during my training.

Thanks in advance!

**Comment by sydney0zq** [comment]
Additionally, I use the validation set of medium 256 version with 2000 cropped images from Places2-standard.

**Comment by windj007** [comment]
> Hi, do you have trained the model to ensure the accuracy is promised?

As stated in the README, we did not yet managed to check reproducibility after code refactoring. However, we did that before the code release - and yes, the results should be reproducible if there wasn't any mistakes during refactoring.

Soon we should have some time and GPUs to actually test reproducibility.


> I have train the lama-fourier with 4 gpus(bs = 10)

Is batch size = 10 denotes sum batch size over 4 gpus? Or this is a batch size for a single gpu (so total batch size is 40)? As both the generator and discriminator use BatchNorms, batch size is critical. When setting total batch size smaller than 20, the quality should degrade significantly.


> Could you please share your training log as a reference.

The exact full log is not currently available, but we will paste it after reproducibility check.

Below are some reference points which are representative of the usual FID/LPIPS dynamics and value ranges. These values are from Big-Lama-Regular and are calculated using a validation set of 2000 images from Places in 512x512 with segmentation-based masks.

| # of iterations | FID   | LPIPS  |
|-----------------|-------|--------|
| 25k             | 17.13 | 0.1088 |
| 100k            | 13.73 | 0.1053 |
| 500k            | 11.32 | 0.0951 |
| 800k            | 10.92 | 0.0924 |

The exact values may vary, but the overal dynamics and the value range should be similar to that.

Before 100k iterations, FID usually fluctuates a lot and sometimes reach 30-50 or even more (but not for these models, which are pretty stable).


> And my FID even reaches to 200 during my training

Please note that FID is heavily dependent on the dataset and its size - and the smaller the dataset the bigger the FID no matter how good the model is. Do you use the mentioned 2000 images to calculate FID? If yes, FID=200 seems to be too big - there might be an error somewhere.

Our pipeline writes intermediate images during training and prints the directory paths it loads data from - did you check them, are they correct?


> And I am interested about one thing, just be curious:

The project is broader than the LaMa paper and is conducted by all the authors of the paper, as a team - and everybodys' contribution is significant. Major development was conducted in a private repo - and the commit history in this one does not reflect actual development process.

**Comment by windj007** [comment]
We're verifying reproducibility with the code from that repo. After ~500k iterations, FID and LPIPS values are similar to the values from my previous comment. Thus, it seems that there is no error in the pipeline.

When the verification is done, we'll upload the full training log and the metrics.

Please check if you're using a correct data.

**Comment by sydney0zq** [comment]
> done

Thanks for your detailed explanation and the solid work. Recently I have no much time to do context alignment with the information you provided, however, I am looking forward to your updating of the full training log and metrics and maybe there are some mistakes in my configuration.

Many thanks for your attention! I will try new experiments without any modification to reproduce your results.

**Comment by sydney0zq** [comment]
@windj007 Recently I have done some further experiments based on your great projects. 

However,  I found the training time is costly.

My hyper-paramters:
- GPU Type: V100 with 32GB * 4
- Dataset: Places512 (cropped to 256 when training)
- BatchSize: 10 Per GPU
- ModelType: Lama-Fourier

It will cost about 12 hours to finish one epoch. Could you please provide some statistics of your training time? And it's better to refer to your full training logs if possible. Thanks in advance!

**Comment by windj007** [comment]
> It will cost about 12 hours to finish one epoch.

What is the epoch size? In our original configs, it is set to 25k iterations per epoch.

If you keep that setting, 12h per epoch seems to be too much. We have approx 6h per epoch on 3xV100 and approx 4-5h per epoch on 4xV100.

Did you check the average load percentage of the GPUs? Maybe they starve data? A clean signal for that is that GPU load drops from 100% to 0 and then raises back to 100 or fluctuates a lot. If this is a case for you, try increasing number of data loading processes or optimize the way you store the dataset.

In our setting, we have the GPUs load between 80 and 100% consistently.
